### PR TITLE
Implement login

### PR DESCRIPTION
**Pull request by julienmachon**
> This is a *temporary* solution, waiting for the IAM documentation and Realm API to be implemented.

### Login Box component (with ability to choose Realm)
screenshot from 2018-11-19 16-19-13
screenshot from 2018-11-19 16-20-05

### Login link in header (only if 1+ realms available)
screenshot from 2018-11-19 16-24-21
screenshot from 2018-11-19 16-24-27

Redux store + react rouder redux connector

### CODE REVIEW

File: src/shared/store/index.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+export default function configureStore(
+  history: History,
+  preloadedState: object = {}
+): Store {
+  const store = createStore(
+    combineReducers({ router: connectRouter(history), ...reducers }),
+    preloadedState,
+    compose(
+      // typeof window === 'object' &&
```
Comment (mdupont) [inline src/shared/store/index.ts]: Commented code

**Comment by mdupont** [comment]
Please use "squash and merge" for this one once approved.

File: package.json [hunk 2]
```diff
@@ -16,14 +16,19 @@
...
   "dependencies": {
     "antd": "^3.10.2",
+    "connected-react-router": "^5.0.1",
     "cookie-parser": "^1.4.3",
     "express": "^4.16.4",
+    "history": "^4.7.2",
     "morgan": "^1.9.1",
-    "react": "^16.7.0-alpha.0",
-    "react-dom": "^16.7.0-alpha.0",
+    "react": "^16.7.0-alpha.2",
```
[the remark this answers is not in the document]
Comment (julienmachon) [reply package.json]: :+1:

File: src/shared/components/Login/Login.test.tsx [hunk 3]
```diff
@@ -1,25 +1,87 @@
...
 
 const authProvider = 'https://keycloack.org';
-const loginComponent = <Login loginURL={authProvider} />;
-const shallowLogin = shallow(loginComponent);
-const fullDOMLogin = mount(loginComponent);
+const realms: Realm[] = [
+  {
+    name: 'BBP',
+    authorizationEndpoint:
+      'https://bbpteam.epfl.ch/auth/realms/BBP/protocol/openid-connect/auth',
```
[the remark this answers is not in the document]
Comment (julienmachon) [reply src/shared/components/Login/Login.test.tsx]: yes. It comes from https://bbpteam.epfl.ch/auth/realms/BBP/.well-known/openid-configuration

File: src/server/index.tsx [hunk 4]
```diff
@@ -64,17 +72,38 @@ app.get('*', (req: express.Request, res: express.Response) => {
...
+
+  // Redux store
+  const store = createStore(memoryHistory, {
+    auth: {
+      accessToken,
+      authenticated: accessToken !== undefined,
+      clientId: process.env.CLIENT_ID || 'nexus-staging',
+      // This is temporary until Realm API is available
+      authorizationEndpoint:
+        'http://staging.nexus.ocp.bbp.epfl.ch/auth/realms/nexus-internal/protocol/openid-connect/auth',
```
[the remark this answers is not in the document]
Comment (julienmachon) [reply src/server/index.tsx]: ah thanks! I wanted this URL yesterday but no one gave it to me :cry: This is temporary anyway, until the API has been implemented

File: src/server/index.tsx [hunk 5]
```diff
@@ -34,19 +36,25 @@ if (process.env.NODE_ENV !== 'production') {
...
+  `${base}/authRedirect`,
+  (req: express.Request, res: express.Response) => {
+    const { access_token } = req.query;
+    res.cookie(
+      cookieName,
+      JSON.stringify({
+        accessToken: access_token,
+      }),
+      {
+        maxAge: 900000,
```
[the remark this answers is not in the document]
Comment (julienmachon) [reply src/server/index.tsx]: I want to set it based on the expire time of the token, but I yet have to implement this

File: docs/authFlow.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+token + realm name is stored in cookie
+when server receives request, it extract the token from the cookie and fetches resources + render app (SSR)
+we are using a cookie and not session so server is stateless.
+we are using a cookie and not local-storage so server gets the token
+
+## 1- No token and no realm
```
[the remark this answers is not in the document]
Comment (julienmachon) [reply docs/authFlow.md]: The API is going to return a list of realms we can authenticate with https://bluebrain.github.io/nexus/docs/api/iam/iam-realms-api.html. They will return the name of the realm and the .well-know file, which will give us the auth/logout links.

**Pull request merged**